### PR TITLE
Facebook external login

### PR DESCRIPTION
**Pull request by jakubfijalkowski**
Adds Facebook to the `LeanCode.ExternalIdentityProviders`. Plus long-needed tests.

There is one place though that I am not sure about. We want to get rid of `LeanCode.Facebook` in the next version and integrate it directly into the provider (as that will be it's only use-case). To make the transition a little bit easier, I register `FacebookModule` along `ExternalIdentityProvidersModule` - this is the part I am not sure about. It will require no changes after the next version but it deviates from what we have up to this point. What do you think?

### CODE REVIEW

File: src/Infrastructure/LeanCode.ExternalIdentityProviders/Facebook/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace LeanCode.ExternalIdentityProviders.Facebook
+{
+    public static class Constants
+    {
+        public const string GrantType = "google";
```
Comment (Saancreed) [inline src/Infrastructure/LeanCode.ExternalIdentityProviders/Facebook/Constants.cs]: That doesn't feel correct.

Comment (lukaszgarstecki) [inline src/Infrastructure/LeanCode.ExternalIdentityProviders/Facebook/Constants.cs]: namespace Facebook and grant type google :thinking:

File: src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs [hunk 2]
```diff
@@ -2,28 +2,77 @@
...
 namespace LeanCode.ExternalIdentityProviders
 {
+    [Flags]
+    public enum Providers
+    {
+        Facebook = 0b001,
+        Apple = 0b010,
+        Google = 0b100,
+
+        None = 0b000,
```
Comment (Saancreed) [inline src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs:19]: Attempting to register this module with `Providers.None` doesn't make much sense so I'd expect it to throw.

**Review by Saancreed — COMMENTED** [review]
I think it's a good time to start v5.1 (removing `LeanCode.Facebook` and keep it unstable for a while) but we could stay on v5.0 and mark everything from the old module as `[Obsolete]`.

**Review by lukaszgarstecki — CHANGES_REQUESTED** [review]

**Comment by Saancreed** [comment]
Also:

> To make the transition a little bit easier, I register `FacebookModule` along `ExternalIdentityProvidersModule` - this is the part I am not sure about.

I feel like there is no point, we should avoid it if possible because it will go away soon. Anyone that uses that module now should migrate asap and if you are about to add Facebook integration just now, go with `ExternalIdentityProvidersModule` only. We will have to deal with two implementations for some time but this way the transition, while requiring some manual intervention, will be cleaner imho.

**Comment by jakubfijalkowski** [comment]
Done, targeting v5.1 now, with `LeanCode.Facebook` removed completely.

**Review by lukaszgarstecki — APPROVED** [review]

File: src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs [hunk 3]
```diff
@@ -2,28 +2,76 @@
...
+    }
+
     public class ExternalIdentityProvidersModule<TUser> : AppModule
         where TUser : IdentityUser<Guid>
     {
+        private readonly Providers configuration = Providers.All;
+
+        public ExternalIdentityProvidersModule(Providers configuration)
+        {
+            if (configuration == Providers.None)
```
Comment (Saancreed) [inline src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs]:
```suggestion
            if ((configuration & Providers.All) == Providers.None)
```
This way we are also guarded against invalid enum values.

File: src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs [hunk 4]
```diff
@@ -2,28 +2,76 @@
...
+        Google = 0b100,
+
+        None = 0b000,
+        All = 0b111,
+    }
+
     public class ExternalIdentityProvidersModule<TUser> : AppModule
         where TUser : IdentityUser<Guid>
     {
+        private readonly Providers configuration = Providers.All;
```
Comment (Saancreed) [inline src/Infrastructure/LeanCode.ExternalIdentityProviders/ExternalIdentityProvidersModule.cs]: Default value is unnecessary now that there is no parameterless constructor.
```suggestion
        private readonly Providers configuration;
```

**Review by Saancreed — APPROVED** [review]
LGTM. Feel free to either implement or ignore my last two suggestions.

**Pull request merged**